### PR TITLE
fix: run array type nil check after item constraints

### PR DESCRIPTION
**Pull request by sheerlox**
Following [my discussion with Zach on the Elixir Forum](https://elixirforum.com/t/array-type-empty-values-cannot-remove-all-empty-values/63330/3), this PR makes sure to run the item constraints before checking for `nil` values in an array type.

The main goal is to address empty strings being converted to `nil` *after* we checked for `nil` values in the array, effectively resulting in `nil` values in the data layer.

# Notes/Questions
- Moving the `nil` check has the side-effect of not running it anymore when `type.custom_apply_constraints_array?()` is true. In the Ash codebase that only impacts `Ash.EmbeddableType`. I don't know what the implications of that change are, so it's worth making sure it's not breaking any existing behavior for anyone.
- I've added a `test/type/array_test.exs` test file, which breaks the 1-to-1 mapping. Should I instead move its content to `test/type/type_test.exs`, given the special array type logic is contained in `ash/type/type.ex`?

# Contributor checklist

- [X] Bug fixes include regression tests
- [ ] Chores
- [ ] Documentation changes
- [ ] Features include unit/acceptance tests
- [ ] Refactoring
- [ ] Update dependencies

### CODE REVIEW

File: lib/ash/type/type.ex [hunk 1]
```diff
@@ -812,31 +812,31 @@ defmodule Ash.Type do
         term
         |> Enum.with_index()
         |> Enum.reduce({[], []}, fn {item, index}, {items, errors} ->
-          if is_nil(item) && not nil_items? do
-            {[item | items], [[message: "no nil values", index: index] | errors]}
+          if type.custom_apply_constraints_array?() do
+            {[item | items], errors}
```
Comment (zachdaniel) [inline lib/ash/type/type.ex:816]: We will need to add a check here for `item` being `nil`, to match the behavior below (as you pointed out)

**Comment by zachdaniel** [comment]
I think the `array_test` is good, as we'll want to test the new constraint of removing empty items that we've talked about, and that gives array-oriented tests a nice place to live 👍

File: lib/ash/type/type.ex [hunk 2]
```diff
@@ -812,31 +812,35 @@ defmodule Ash.Type do
         term
         |> Enum.with_index()
         |> Enum.reduce({[], []}, fn {item, index}, {items, errors} ->
-          if is_nil(item) && not nil_items? do
-            {[item | items], [[message: "no nil values", index: index] | errors]}
-          else
-            if type.custom_apply_constraints_array?() do
-              {[item | items], errors}
+          if type.custom_apply_constraints_array?() do
+            if is_nil(item) && not nil_items? do
```
Comment (sheerlox) [inline lib/ash/type/type.ex:816]: Is the check duplication OK, or is there a better/cleaner way to do this? Still kinda new to Elixir so I don't master the best practices yet :sweat_smile:

Comment (sheerlox) [reply lib/ash/type/type.ex:816]: Well I'm working on the next PR (`remove_nil_items?`) and created a private function in there to centralize the duplicated code :+1:

Comment (zachdaniel) [reply lib/ash/type/type.ex:816]: A little duplication never hurt anybody 😆

**Comment by zachdaniel** [comment]
Will merge once build passes

**Review by zachdaniel — APPROVED** [review]

**Comment by zachdaniel** [comment]
🚀 Thank you for your contribution! 🚀

**Pull request merged**